tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs

```diff
@@ -281,6 +291,233 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
+
+			if (type.HasGenericParameters) {
+				ctor = type.CreateMethodReferenceOnGenericType (ctor, type.GenericParameters.ToArray ());
+			}
+
+			// return new TypeA (nativeHandle); // for NativeHandle ctor
+			// return new TypeA ((IntPtr) nativeHandle); // for IntPtr ctor
+			il.Emit (OpCodes.Ldarg, nativeHandleParameter);
+			if (ctor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
+				il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
+			il.Emit (OpCodes.Newobj, ctor);
+			il.Emit (OpCodes.Ret);
+
+			Annotations.Mark (createInstanceMethod);
+		}
+
+		void ImplementConstructINativeObjectFactoryMethod (TypeDefinition type, MethodReference? ctor)
+		{
+			// skip creating the factory for NSObject itself
+			if (type.Is ("Foundation", "NSObject"))
+				return;
+
+			// If the type is a subclass of NSObject, we prefer the NSObject "IntPtr" constructor
+			var nsobjectConstructor = type.IsNSObject (DerivedLinkContext) ? FindNSObjectConstructor (type) : null;
+			if (nsobjectConstructor is null && ctor is null)
+				return;
+
+			var createInstanceMethod = type.AddMethod ("ConstructINativeObject", MethodAttributes.Public | MethodAttributes.Static | MethodAttributes.NewSlot | MethodAttributes.HideBySig, abr.ObjCRuntime_INativeObject);
+			var nativeHandleParameter = createInstanceMethod.AddParameter ("nativeHandle", abr.ObjCRuntime_NativeHandle);
+			var ownsParameter = createInstanceMethod.AddParameter ("owns", abr.System_Boolean);
+			createInstanceMethod.Overrides.Add (abr.INativeObject_ConstructINativeObject);
+			var body = createInstanceMethod.CreateBody (out var il);
+
+			if (nsobjectConstructor is not null) {
+				// var instance = new TypeA (nativeHandle);
+				// // alternatively with a cast: new TypeA ((IntPtr) nativeHandle);
+				// if (instance is not null && owns)
+				//     Runtime.TryReleaseINativeObject (instance);
+				// return instance;
+
+				if (type.HasGenericParameters) {
+					nsobjectConstructor = type.CreateMethodReferenceOnGenericType (nsobjectConstructor, type.GenericParameters.ToArray ());
+				}
+
+				var instanceVariable = body.AddVariable (abr.Foundation_NSObject);
+				il.Emit (OpCodes.Ldarg, nativeHandleParameter);
+				if (nsobjectConstructor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
+					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
+				il.Emit (OpCodes.Newobj, nsobjectConstructor);
+				il.Emit (OpCodes.Stloc, instanceVariable);
+
+				var falseTarget = il.Create (OpCodes.Nop);
+				il.Emit (OpCodes.Ldloc, instanceVariable);
+				il.Emit (OpCodes.Ldnull);
+				il.Emit (OpCodes.Cgt_Un);
+				il.Emit (OpCodes.Ldarg, ownsParameter);
+				il.Emit (OpCodes.And);
+				il.Emit (OpCodes.Brfalse_S, falseTarget);
+
+				il.Emit (OpCodes.Ldloc, instanceVariable);
+				il.Emit (OpCodes.Call, abr.Runtime_TryReleaseINativeObject);
+
+				il.Append (falseTarget);
+
+				il.Emit (OpCodes.Ldloc, instanceVariable);
+				il.Emit (OpCodes.Ret);
```

Comment:
You don't actually need the variable:

You don't actually need the variable:

```suggestion
				il.Emit (OpCodes.Ldarg, nativeHandleParameter);
				if (nsobjectConstructor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
				il.Emit (OpCodes.Newobj, nsobjectConstructor);

				var falseTarget = il.Create (OpCodes.Nop);
				il.Emit (OpCodes.Ldloc, instanceVariable);
				il.Emit (OpCodes.Ldnull);
				il.Emit (OpCodes.Cgt_Un);
				il.Emit (OpCodes.Ldarg, ownsParameter);
				il.Emit (OpCodes.And);
				il.Emit (OpCodes.Brfalse_S, falseTarget);

				il.Emit (OpCodes.Dup);
				il.Emit (OpCodes.Call, abr.Runtime_TryReleaseINativeObject);

				il.Append (falseTarget);

				il.Emit (OpCodes.Ret);
```